includes/Data/Patterns.php

```diff
@@ -203,27 +236,69 @@ public static function get_meta_from_pattern_slug( $pattern_slug ) {
 	}
 
 	/**
-	 * Retrieve pattern from slug.
+	 * Get the fallback slug for a given slug.
 	 *
-	 * @param string $pattern_slug Pattern Slug Data
+	 * @param string $pattern_slug The given slug.
+	 * @return string|false
+	 */
+	public static function get_fallback_from_slug( $pattern_slug ) {
+		$theme_pattern = explode( '/', $pattern_slug );
+		if ( ! isset( $theme_pattern[0] ) || ! isset( $theme_pattern[1] ) ) {
+			return false;
+		}
+		$fallbacks = self::get_fallbacks();
+		return isset( $fallbacks[ $theme_pattern [0] ][ $theme_pattern [1] ] )
+		? $fallbacks[ $theme_pattern [0] ][ $theme_pattern [1] ]
+		: false;
+	}
+
+	/**
+	 * Fetches a pattern from the WP_Block_Patterns_Registry.
 	 *
-	 * @return array|boolean
+	 * @param string $pattern_slug The full slug of the pattern.
+	 * @return array
 	 */
-	public static function get_pattern_from_slug( $pattern_slug ) {
+	public static function get_pattern_from_block_patterns_registry( $pattern_slug ) {
 		$block_patterns_registry = \WP_Block_Patterns_Registry::get_instance();
 		if ( $block_patterns_registry->is_registered( $pattern_slug ) ) {
 			$pattern = $block_patterns_registry->get_registered( $pattern_slug );
 			return array(
-				'title'   => $pattern['title'],
-				'content' => self::cleanup_wp_grammar( $pattern['content'] ),
-				'name'    => $pattern['name'],
-				'meta'    => self::get_meta_from_pattern_slug( $pattern_slug ),
+				'slug'       => $pattern_slug,
+				'title'      => $pattern['title'],
+				'content'    => self::cleanup_wp_grammar( $pattern['content'] ),
+				'name'       => $pattern['name'],
+				'meta'       => self::get_meta_from_pattern_slug( $pattern_slug ),
+				'categories' => $pattern['categories'],
 			);
 		}
-
 		return false;
 	}
 
+	/**
+	 * Retrieve pattern from slug.
+	 *
+	 * @param string $pattern_slug Pattern Slug Data
+	 *
+	 * @return array|boolean
+	 */
+	public static function get_pattern_from_slug( $pattern_slug ) {
```

Comment:
In most cases, we are retrieving a single pattern in Onboarding, this also gives us better control over fallbacks and child theme generation. Even the homepage-styles step retrieves 2 patterns since `home-1` should have already been cached. Seems like an overkill to retrieve all homepages and then filter the one that is needed at that point (do we eventually do that for headers/footers as well?). The mass retrieval using category exists in the patterns module but is not something we need in Onboarding. An endpoint that accepts a list of patterns, is something we can think of if we start using headers, footers and page sections from WonderBlocks.